common/kafka/kafka_consumer_holder.cpp

```diff
@@ -0,0 +1,448 @@
+                  << ", offSet: " << topic_partition.get()->offset()
+                  << std::endl << std::flush;
+
+
+        if (!KafkaSeekWithRetry(kafkaConsumer_.get(), *topic_partition)) {
+          return false;
+        }
+      } else {
+        LOG(ERROR) << "Incorrect state for consumed offset for topic: " << consumedOffset.first.first
+                   << ", partition: " << consumedOffset.first.second
+                   << ", timestamp: " << consumedOffset.second.timestamp
+                   << ", offSet: " << consumedOffset.second.offset
+                   << std::endl << std::flush;
+        return false;
+      }
+
+      LOG(INFO) << "Successfuly seeked to offset for topic: " << consumedOffset.first.first
+                << ", partition: " << consumedOffset.first.second
+                << ", timestamp: " << consumedOffset.second.timestamp
+                << ", offSet: " << topic_partition.get()->offset()
+                << std::endl << std::flush;
+    }
+    return true;
+  }
+
+  virtual void resetInstance(const TopicPartitionToValueMap<ConsumedOffset>* consumedOffsets) override {
+    syncPreviousInstance();
+    /*
+     * Before creating a new KafkaConsumer instance, make sure old one is closed
+     * and shutdown.
+     */
+    bool done = false;
+    do {
+      closePreviousInstance();
+
+      recreateInstance();
+
+      std::this_thread::sleep_for(
+        std::chrono::milliseconds(1000));
+
+      /**
+       * If there are no consumedOffsets, return;
+       */
+      if (consumedOffsets == nullptr) {
+        return;
+      }
+
+      if (!assignPartitions()) {
+        continue;
+      }
+
+      if (!seekToRelevantOffsets(consumedOffsets)) {
+        continue;
+      }
+      done = true;
+    } while (!done);
+  }
+
+  virtual void close() override {
+    if (kafkaConsumer_) {
+      kafkaConsumer_->close();
+    }
+  }
+};
+
+class RdKafkaConsumerHolderCached : virtual public RdKafkaConsumerHolder {
+private:
+  const std::shared_ptr<RdKafka::KafkaConsumer> kafkaConsumer_;
+public:
+  RdKafkaConsumerHolderCached(std::shared_ptr<RdKafka::KafkaConsumer> kafkaConsumer)
+    : kafkaConsumer_(std::move(kafkaConsumer)) {}
+
+  virtual std::shared_ptr<RdKafka::KafkaConsumer> getInstance() override {
+    return kafkaConsumer_;
+  }
+
+  virtual void resetInstance(const TopicPartitionToValueMap<ConsumedOffset>* consumedOffsets) override {
+    // doNothing();
+  }
+
+  virtual void close() override {
+    if (kafkaConsumer_) {
+      kafkaConsumer_->close();
+    }
+  }
+};
+
+RdKafkaConsumerHolder* RdKafkaConsumerHolderFactory::createInstance(
+  const std::unordered_set<uint32_t>& partition_ids,
+  const std::string& broker_list,
+  const std::unordered_set<std::string>& topic_names,
+  const std::string& group_id,
+  const std::string& kafka_consumer_type) {
+  return new RdKafkaConsumerHolderRenewable(partition_ids, broker_list, topic_names, group_id, kafka_consumer_type);
```

Comment:
So will we be using a `RdKafkaConsumerHolderRenewable` consumer for PII use cases and `RdKafkaConsumerHolderCached` for non-PII?

Reply:
No. RdKafkaConsumerHolderCached is cache of already constructed RdKafkaConsumer instance... It doesn't know how to re-construct it.... So, it doesn't reset anything... This is used in test.

In production scenario, the RdKafkaConsumer is constructed from broker file providing broker instances.. RdKafkaConsumerHolderRenewable holds on to information required to re-construct RdKafkaConsumer. It is used both when no file is watched (where resetInstance is never called, and when a file is watched for changes and triggers resetInstance from KafkaConsumer class.)